taichi/transforms/whole_kernel_cse.cpp

```diff
@@ -0,0 +1,69 @@
+#include "taichi/ir/ir.h"
+
+TLANG_NAMESPACE_BEGIN
+
+// Whole Kernel Common Subexpression Elimination
+class WholeKernelCSE : public BasicStmtVisitor {
+ public:
+  using BasicStmtVisitor::visit;
+
+  WholeKernelCSE() {}
+
+  void visit(IfStmt *if_stmt) override {
+    // Move common statements at the beginning or the end of both branches
+    // outside.
+    if (if_stmt->true_statements && if_stmt->false_statements) {
+      auto block = if_stmt->parent;
+      int current_stmt_id = block->locate(if_stmt);
+      std::cout << "wholekernel cse if" << std::endl;
+      irpass::print(block->statements[current_stmt_id - 1].get());
+      irpass::print(if_stmt);
+      auto &true_clause = if_stmt->true_statements;
+      auto &false_clause = if_stmt->false_statements;
+      if (irpass::same_statements(true_clause->statements[0].get(),
+                                  false_clause->statements[0].get())) {
+        std::cout << "same!";
+
+        irpass::print(true_clause->statements[0].get());
+        irpass::print(false_clause->statements[0].get());
+        auto common_stmt = std::move(true_clause->statements[0]);
+        irpass::replace_all_usages_with(false_clause.get(),
+                                        false_clause->statements[0].get(),
+                                        common_stmt.get());
+        true_clause->statements.erase(true_clause->statements.begin());
```

Comment:
This is kind of hacky. I want to move one of the common statements outside (so there is no need to move it to the trash bin), and delete the other statement.

Reply:
Wait, please always keep deleted statements in the trash bin otherwise `==` won't work for `Stmt *`, since the pointed addresses are no longer unique.

Reply:
How to copy a statement outside the clause then?

Reply:
Actually `Block::erase` will simply move the statement to the trash bin instead of **deleting** it and let the memory allocator reclaim the memory. So I think you should be fine here :-)

(Sorry I thought you were deleting the statement manually.)

Reply:
Oh I should've used the word erase. I was saying that I used `true_clause->statements.erase` but not `true_clause->erase` to avoid moving a `nullptr` to the trash bin.

Reply:
I see. To prevent this from happening again, maybe we should make `statements` private?

Reply:
But how to move a statement outside if `statements` is private? Shall we add a function to move the `unique_ptr` outside the clause rather than to the trash bin?

Reply:
It sounds to me that this can't be done with just a few lines of code. I need some time to think about a systematic solution. For now, let's just be cautious.

Reply:
If I understand it correctly, I have implemented the function `Block::extract` to move a `unique_ptr<Stmt>` outside `statements`. The behavior should be the same as the previous commit in this PR.

Reply:
Which commit are you talking about?

Reply:
I mean https://github.com/taichi-dev/taichi/pull/727/commits/0e4d3a66078993067f5fc34d42545db790efa59b should not change the behavior of the code.

Reply:
I think you are right.